### DISCUSSION TITLE
New to all this, little help?

### DISCUSSION THREAD

**Discussion by thehijacker** [Tyrrrz/CliWrap#227] [Q&A]
I have an old application that fires multiple background processes written in .NET 4.5. Now migrated it to 7.0 and wish to replace the ugly code I had with CliWrap because I am sure it is proper and modern way to do it. But I am stuck with few problems that I had solved with old code, but can not figure out how to do them now.

1. I was using new Process and feed it's StandardOutput and StandardError to a function for further processing. Here I can only figure out how to pass the output and error to console. How to push this to external function for live processing?

2. It is a console application but even a simple sample doesn't work for me. Under main() I fire new background thread that runs a php server and closes it after 5 seconds. Process exits but the task never finishes. Sample code I am using:

```
Task.Run(async () =>
{
    using var cts = new CancellationTokenSource();
    cts.CancelAfter(TimeSpan.FromSeconds(5));

    var task = Cli.Wrap("/usr/bin/php")
        .WithArguments(["-S", "172.21.63.11:9901", "-t", "/home/t2admin/T2RelayMgmt"])
        .WithStandardErrorPipe(PipeTarget.ToDelegate(Console.WriteLine))
        .WithStandardOutputPipe(PipeTarget.ToDelegate(Console.WriteLine))
        .ExecuteAsync(cts.Token);

    Logger.Info("Running. Proces ID: {0}", task.ProcessId);

    Logger.Debug("Before await");
    await task;
    Logger.Debug("After await");
});
```

3. I will also need to be able to kill process. Using the cancelation tokens if I read corretly. What is the proper way to handle this if I have lets say 16 processes and wish to kill specific one?

I wish there were more samples in this project to use a guidence.

Thank you.

**Comment by Tyrrrz** [comment]
> 1. I was using new Process and feed it's StandardOutput and StandardError to a function for further processing. Here I can only figure out how to pass the output and error to console. How to push this to external function for live processing?

You can put anything else inside your `PipeTarget.ToDelegate(...)` delegate.

> 2\. It is a console application but even a simple sample doesn't work for me. Under main() I fire new background thread that runs a php server and closes it after 5 seconds. Process exits but the task never finishes. Sample code I am using:

Make sure you `await` your background task.

> I will also need to be able to kill process. Using the cancelation tokens if I read corretly. What is the proper way to handle this if I have lets say 16 processes and wish to kill specific one?

You issue separate cancellation tokens for each one.

**Reply by thehijacker** [reply]
> > 1. I was using new Process and feed it's StandardOutput and StandardError to a function for further processing. Here I can only figure out how to pass the output and error to console. How to push this to external function for live processing?
> 
> You can put anything else inside your `PipeTarget.ToDelegate(...)` delegate.

Can not make it work. How to get function ProcessOutput function as delegate?

```
RunProcess(
          c,
          executable,
          parameters,
          s => { ProcessOutput(c, s); },
          s => { ProcessOutput(c, s); }
      );
```

> > 2. It is a console application but even a simple sample doesn't work for me. Under main() I fire new background thread that runs a php server and closes it after 5 seconds. Process exits but the task never finishes. Sample code I am using:
> 
> Make sure you `await` your background task.

It is awaited, but it still doesn't finish.

`await task;`

> > I will also need to be able to kill process. Using the cancelation tokens if I read corretly. What is the proper way to handle this if I have lets say 16 processes and wish to kill specific one?
> 
> You issue separate cancellation tokens for each one.

Got it. When I am that far :).

Thank you.
